### PR TITLE
PID updated. to focus on mu/pi separation for relatively low momentum…

### PR DESCRIPTION
**Pull request by MasakazuKurata**
I tried to push updated PID again. Please check that.

BEGINRELEASENOTES
- Focus on low momentum mu/pi separation
- Correct the corresponding change for createPDF processor
 
- update: correct bad lines:
  1. use DD4hepAPI to get b field
  2. correct array initialization
  3. prevent memory leak when using ROOT

ENDRELEASENOTES

### CODE REVIEW

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 1]
```diff
@@ -381,10 +417,13 @@ void CreatePDFProcessor::init() {
...
   _myPID = new LikelihoodPID(pars);
 
+  //get B field
+  _bfield = marlin::Global::GEAR->getBField().at(0.0,0.0,0.0)[2];
+
   printParameters();  
 }
 
-void CreatePDFProcessor::processRunHeader( LCRunHeader*  /*run*/) { 
+void CreatePDFProcessor::processRunHeader( LCRunHeader* run) {
```
Comment (rete) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: Put back the comment please

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 2]
```diff
@@ -489,7 +537,7 @@ void CreatePDFProcessor::end() {
   _fpdf = new TFile(_filename.c_str(), "RECREATE");
```
Comment (rete) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: Memory leak here with _fpdf, but also with _myPID that should both deleted in the `end()` method

**Comment by rete** [comment]
Please follow the guideline for your PR message. I will validate it after

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 3]
```diff
@@ -2,8 +2,9 @@
 #include <vector>
 
 #include <marlin/Global.h>
-#include "gear/GEAR.h"
-#include "gearimpl/ConstantBField.h"
+#include "DD4hep/Detector.h"
+#include "DDSurfaces/Vector3D.h"
+#include "DD4hep/DD4hepUnits.h"
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: Erase GEAR and use DD4hepAPI

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 4]
```diff
@@ -417,13 +418,16 @@ void CreatePDFProcessor::init() {
 
   _myPID = new LikelihoodPID(pars);
 
-  //get B field
-  _bfield = marlin::Global::GEAR->getBField().at(0.0,0.0,0.0)[2];
+  //get B field                                          
+  double bfield[3]={};
+  dd4hep::Detector& lcdd = dd4hep::Detector::getInstance();
+  lcdd.field().magneticField({0.0,0.0,0.0}, bfield);
+  _bfield = (float)bfield[2]/dd4hep::tesla;
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: Get B field from hh4hep API

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 5]
```diff
@@ -417,13 +418,16 @@ void CreatePDFProcessor::init() {
...
 
   printParameters();  
 }
 
-void CreatePDFProcessor::processRunHeader( LCRunHeader* run) { 
+void CreatePDFProcessor::processRunHeader( LCRunHeader* /*run*/) {
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: comment out

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 6]
```diff
@@ -528,7 +532,7 @@ void CreatePDFProcessor::processEvent( LCEvent * evt ) {
   }  
 }
 
-void CreatePDFProcessor::check( LCEvent * evt ) { 
+void CreatePDFProcessor::check( LCEvent * /*evt*/ ) {
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: comment out

File: Analysis/PIDTools/src/CreatePDFProcessor.cc [hunk 7]
```diff
@@ -543,6 +547,9 @@ void CreatePDFProcessor::end() {
   }
 
   _fpdf->Close();
+
+  delete _fpdf;
+  delete _myPID;
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/CreatePDFProcessor.cc]: delete pointers to prevent memory leak

File: Analysis/PIDTools/src/LikelihoodPID.cc [hunk 8]
```diff
@@ -23,9 +23,9 @@ risk-minimization and MAP
 #include <TFile.h>
 #include <TH1F.h>
 
-#include "marlin/Global.h"
-#include "gear/GEAR.h"
-#include "gearimpl/ConstantBField.h"
+#include "DD4hep/Detector.h"
+#include "DDSurfaces/Vector3D.h"
+#include "DD4hep/DD4hepUnits.h"
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/LikelihoodPID.cc]: use DD4hep API instead of GEAR

File: Analysis/PIDTools/src/LikelihoodPID.cc [hunk 9]
```diff
@@ -65,7 +65,10 @@ LikelihoodPID::LikelihoodPID(double *pars){
   pmass=0.938272;
 
   //get B field
-  _bfield = marlin::Global::GEAR->getBField().at(0.0,0.0,0.0)[2];
+  double bfield[3]={};
+  dd4hep::Detector& lcdd = dd4hep::Detector::getInstance();
+  lcdd.field().magneticField({0.0,0.0,0.0}, bfield);
+  _bfield = (float)bfield[2]/dd4hep::tesla;
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/LikelihoodPID.cc]: get B field from DD4hep API

File: Analysis/PIDTools/src/LikelihoodPID.cc [hunk 10]
```diff
@@ -256,7 +259,10 @@ LikelihoodPID::LikelihoodPID(string fname, double *pars, std::vector<float> cost
   }
 
   //get z component of B field
-  _bfield = marlin::Global::GEAR->getBField().at(0.0,0.0,0.0)[2];
+  double bfield[3]={};
+  dd4hep::Detector& lcdd = dd4hep::Detector::getInstance();
+  lcdd.field().magneticField({0.0,0.0,0.0}, bfield);
+  _bfield = (float)bfield[2]/dd4hep::tesla;
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/LikelihoodPID.cc]: get B field from DD4hep API

File: Analysis/PIDTools/src/LikelihoodPID.cc [hunk 11]
```diff
@@ -500,7 +506,7 @@ int LikelihoodPID::Class_electron(TLorentzVector pp, EVENT::Track* trk, EVENT::C
   if(cluvec.size()!=0) shapes=cluvec[0]->getShape();
 
   //get variables
-  double var[11]={0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0}; //11
+  double var[11]={}; //11
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/LikelihoodPID.cc]: Change initialization

File: Analysis/PIDTools/src/LikelihoodPID.cc [hunk 12]
```diff
@@ -664,7 +670,7 @@ int LikelihoodPID::Class_muon(TLorentzVector pp, EVENT::Track* trk, EVENT::Clust
   if(cluvec.size()!=0) shapes=cluvec[0]->getShape();
 
   //get variables
-  double var[17]={0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0}; //17 variales used(is it OK?)
+  double var[17]={}; //17 variales used(is it OK?)
```
Comment (MasakazuKurata) [inline Analysis/PIDTools/src/LikelihoodPID.cc]: Change initialization

**Comment by rete** [comment]
The pull request template says this : 
```
BEGINRELEASENOTES
- Thank you for writing the text to appear in the release notes. It will show up
  exactly as it appears between the two bold lines
- ...

ENDRELEASENOTES
```
Please, edit your initial PR message to match this one.

**Pull request merged**